build_runner_core/lib/src/asset/batch.dart

```diff
@@ -0,0 +1,191 @@
+/// possible that other running tools such as an analysis server will have to
+/// re-analyze incomplete states multiple times.
+/// By storing pending outputs in memory first and then committing them at the
+/// end of the build, we have a better view over that needs to happen.
+///
+/// The default [IOEnvironment] uses readers and writes that are batch-aware
+/// outside of low-memory mode.
+@internal
+final class FileSystemWriteBatch {
+  final Map<AssetId, _PendingFileState> _pendingWrites = {};
+
+  FileSystemWriteBatch._();
+
+  Future<void> _completeWrites(RunnerAssetWriter writer) async {
+    await Future.wait(_pendingWrites.entries.map((e) async {
+      if (e.value.content case final content?) {
+        await writer.writeAsBytes(e.key, content);
+      } else {
+        await writer.delete(e.key);
+      }
+    }));
+
+    _pendingWrites.clear();
+  }
+
+  static FileSystemWriteBatch? get current {
+    return Zone.current[_key] as FileSystemWriteBatch?;
+  }
+
+  static final _key = Object();
+}
+
+final class _PendingFileState {
+  final List<int>? content;
+
+  const _PendingFileState(this.content);
+
+  bool get isDeleted => content == null;
+}
+
+@internal
+final class BatchAwareReader extends AssetReader
+    implements RunnerAssetReader, PathProvidingAssetReader {
+  final RunnerAssetReader _inner;
+  final PathProvidingAssetReader _innerPathProviding;
+
+  BatchAwareReader(this._inner, this._innerPathProviding);
+
+  _PendingFileState? _stateFor(AssetId id) {
+    if (FileSystemWriteBatch.current case final batch?) {
+      return batch._pendingWrites[id];
+    } else {
+      return null;
+    }
+  }
+
+  @override
+  Future<bool> canRead(AssetId id) async {
+    if (_stateFor(id) case final state?) {
+      return !state.isDeleted;
+    } else {
+      return await _inner.canRead(id);
+    }
+  }
+
+  @override
+  Stream<AssetId> findAssets(Glob glob, {String? package}) {
+    return _inner
+        .findAssets(glob, package: package)
+        .where((asset) => _stateFor(asset)?.isDeleted != true);
+  }
+
+  @override
+  String pathTo(AssetId id) {
+    return _innerPathProviding.pathTo(id);
+  }
+
+  @override
+  Future<List<int>> readAsBytes(AssetId id) async {
+    if (_stateFor(id) case final state?) {
+      if (state.isDeleted) {
+        throw AssetNotFoundException(id);
+      } else {
+        return state.content!;
+      }
+    } else {
+      return await _inner.readAsBytes(id);
+    }
+  }
+
+  @override
+  Future<String> readAsString(AssetId id, {Encoding encoding = utf8}) async {
+    final bytes = await readAsBytes(id);
+    return encoding.decode(bytes);
+  }
+}
+
+@internal
+final class BatchAwareWriter extends RunnerAssetWriter {
+  final RunnerAssetWriter _inner;
+
+  BatchAwareWriter(this._inner);
+
+  FileSystemWriteBatch? _batchFor(AssetId id) {
+    // Don't batch writes for hidden files unlikely to be consumed by other
+    // tools.
+    if (id.pathSegments case ['.dart_tool', ...]) {
```

Comment:
`id.pathSegments` is fairly expensive to do a lot, I would just check for a string prefix.

```suggestion
    if (id.path.startsWith('.dart_tool/') {
```

However, I am curious what prompted you to add this? It feels like it is hard coding in an assumption  that might not hold later on.

Reply:
>However, I am curious what prompted you to add this? It feels like it is hard coding in an assumption that might not hold later on.

My idea was to reduce the size of the in-memory batch by not including files that are unlikely to be observed by other tools directly (like hidden assets in `.dart_tool/build/generated`).
I can remove that if we want that behavior, or otherwise add a comment explaining the motivation for that check.

Reply:
I think it probably just isn't worth doing - generally these build-to-cache files are going to eventually get read back by something else (and subsequently cached) anyways, so there won't actually be any extra memory used, and we can avoid re-reading the file this way.

Also, I believe the analyzer _does_ look in the build cache for files.